internal/firmware/revision.go

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package firmware
+
+import (
+	"fmt"
+	"strconv"
+)
+
+type Revision struct {
+	N int
+}
+
+func (r Revision) String() string {
+	if r.N == 0 {
+		return "unset"
+	}
+	if r.N < 0 {
+		return fmt.Sprintf("x%d", -r.N)
+	}
+	return strconv.Itoa(r.N)
+}
+
+func (r Revision) Unset() bool {
+	return r.N == 0
+}
+
+func (r Revision) Local() bool {
+	return r.N < 0
+}
+
+func (r Revision) Store() bool {
+	return r.N > 0
+}
+
+func (r Revision) MarshalJSON() ([]byte, error) {
+	return []byte(`"` + r.String() + `"`), nil
+}
+
+func (r *Revision) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	return r.UnmarshalJSON([]byte(`"` + s + `"`))
+}
+
+func (r Revision) MarshalYAML() (interface{}, error) {
+	return r.String(), nil
+}
+
+func (r *Revision) UnmarshalJSON(data []byte) error {
+	if len(data) > 0 && data[0] == '"' && data[len(data)-1] == '"' {
+		parsed, err := ParseRevision(string(data[1 : len(data)-1]))
+		if err == nil {
+			*r = parsed
+			return nil
+		}
+	} else {
+		n, err := strconv.ParseInt(string(data), 10, 64)
+		if err == nil {
+			r.N = int(n)
+			return nil
+		}
+	}
+	return fmt.Errorf("invalid firmware revision: %q", data)
+}
+
+// ParseRevisions returns the representation in r as a revision.
+// See R for a function more suitable for hardcoded revisions.
+func ParseRevision(s string) (Revision, error) {
+	if s == "unset" {
+		return Revision{}, nil
+	}
+	if s != "" && s[0] == 'x' {
+		i, err := strconv.Atoi(s[1:])
+		if err == nil && i > 0 {
+			return Revision{-i}, nil
+		}
+	}
+	i, err := strconv.Atoi(s)
+	if err == nil && i > 0 {
+		return Revision{i}, nil
+	}
+	return Revision{}, fmt.Errorf("invalid firmware revision: %#v", s)
```

Comment:
err is not fully handled here, nor is "why" the firmware revision is invalid (i.e. -ve revision number)

Reply:
Hi Harry. I am going to try to cover the last 3 comments in one here:

Just a reminder that the revision code is taken verbatim from snapd, and has been in use forever. I keep the two versions in sync as both KernOS and snapd needs it for very similar reasons.

Some details to understand how the system works. 

```revision > 0``` => positive revision number which indicates the revision was assigned by a product store authority. 
```revision < 0``` => negative revision, but the string representation replaces the minus with ```x```.

The CLI interface and other sources can use R() to convert either a int or string format of a revision.

For the string case, if its not:

local revision: x1, x5, x122 etc..
store revision: 1, 5, 122 etc...

Then its invalid.